README.md
[![Build Status](https://travis-ci.org/KevinFasusi/supplychainpy.svg?branch=master)](https://travis-ci.org/KevinFasusi/supplychainpy?branch=master)
[![Documentation Status](https://readthedocs.org/projects/supplychainpy/badge/?version=latest)](http://supplychainpy.readthedocs.org/en/latest/?badge=latest)
[![Coverage Status](https://coveralls.io/repos/github/KevinFasusi/supplychainpy/badge.svg?branch=master)](https://coveralls.io/github/KevinFasusi/supplychainpy?branch=master)
[![PyPI version](https://badge.fury.io/py/supplychainpy.svg)](https://badge.fury.io/py/supplychainpy)
[![Requirements Status](https://requires.io/github/KevinFasusi/supplychainpy/requirements.svg?branch=master)](https://requires.io/github/KevinFasusi/supplychainpy/requirements/?branch=master)
![Supplychainpy logo](https://github.com/KevinFasusi/supplychainpy/blob/master/supplychainpy/reporting/static/images/logo-trn.png)

# Supplychainpy

Supplychainpy is a Python library for supply chain analysis, modelling and simulation. The library assists a workflow that is reliant on Excel and VBA.
Quite often Demand Planners, Buyers, Supply Chain Analysts and BI Analysts have to create their tools in Microsoft Excel for one reason or another. 
In a perfect world, the ERP/MRP system would be sufficient, but this is not always the case. Some issues include:

- Building reports and performing demand forecasts or planning inventory can become repetitive. 
- Visualisation tools are often platform dependent. 

It is the aim of this library to:

- Alleviate the reliance on Excel.
- Free up the analysts from some of the most mundane tasks of building tools.
- Provide a space for innovation and implementing advancements in the domain.
- Leverage the robust and extensible Python ecosystem.

The library is currently in early stages of development, so not ready for use in production. For quick exploration, please see the **Quick Guide** below.

## Quick Install

The easiest way to install supplychainpy is via pip: `pip install supplychainpy`.

An alternative is to clone the repository:

1. Run the command: `python setup.py sdist` 

2. Navigate to the `dist` folder.

3. Run the command: `pip install supplychainpy-0.0.4.tar.gz`


## Dependencies

- Numpy
- Pandas
- Flask
- Flask-Restful
- Flask-Restless
- Flask-Script
- Flask-SqlAlchemy
- Flask-Uploads
- Flask-WTF
- Scipy
- SqlAlchemy
- TextBlob


## Optional Dependencies

- matplotlib
- xlwings
- openpyxl


## Python Version

- python 3.5

## Quick Guide

Below is a quick example using a [sample data file](https://github.com/KevinFasusi/supplychainpy/blob/master/supplychainpy/sample_data/complete_dataset_small.csv). 

```python
    
from supplychainpy.model_inventory import analyse
from supplychainpy.sample_data.config import ABS_FILE_PATH
from decimal import Decimal

analysed_data = analyse(
						file_path=ABS_FILE_PATH['COMPLETE_CSV_SM'],
						z_value=Decimal(1.28),
                        reorder_cost=Decimal(400),
                        retail_price=Decimal(455),
                        file_type='csv',
                        currency='USD'
                        )
analysis = [demand.orders_summary() for demand in analysed_data] 
```

output:

```
{'reorder_level': '4069', 'orders': {'demand': ('1509', '1855', '2665', '1841', '1231', '2598', '1988', '1988', '2927', '2707', '731', '2598')}, 'total_orders': '24638', 'economic_order_quantity': '44', 'sku': 'KR202-209', 'unit_cost': '1001', 'revenue': '123190000', 'quantity_on_hand': '1003', 'shortages': '5969', 'excess_stock': '0', 'average_orders': '2053.1667', 'standard_deviation': '644', 'reorder_quantity': '13', 'safety_stock': '1165', 'demand_variability': '0.314', 'ABC_XYZ_Classification': 'BY', 'economic_order_variable_cost': '15708.41', 'currency': 'USD'} ...
```


Alternatively using a Pandas `DataFrame`:

```python

from supplychainpy.model_inventory import analyse
from supplychainpy.sample_data.config import ABS_FILE_PATH
from decimal import Decimal
import pandas as pd

raw_df = pd.read_csv(ABS_FILE_PATH['COMPLETE_CSV_SM'])
analyse_kv = dict(
					df=raw_df,
        			start=1,
        			interval_length=12,
        			interval_type='months',
        			z_value=Decimal(1.28),
        			reorder_cost=Decimal(400),
			        retail_price=Decimal(455),
        			file_type='csv',
        			currency='USD'
        			)

analysis_df = analyse(**analyse_kv)

```


Further examples please refer to the jupyter notebooks [here](https://github.com/KevinFasusi/supplychainpy_notebooks).
For more detailed coverage of the api please see the [documentation](http://supplychainpy.readthedocs.org/).

## New Reporting Feature

The reports include a dashboard, raw analysis, a recommendations feed and SKU level analysis with forecast:

![reports](https://github.com/supplybi/supplybi.github.io/blob/master/images/reports.gif)


Launch reports can be achieved from the command line by:

```
    supplychainpy data.csv -a -loc absolute/path/to/current/directory -l
```

Other optional arguments include the host (--host default: 127.0.0.1 ) and port (-p default: 5000) arguments. Setting the host and ports allows the `-l` arguments can be replaced by the `-lx`. The `-l` arguments launch a small intermediary GUI for setting the port before launching the reports in a web browser. The `-lx` argument start the reporting process but does not launch a GUI or a browser window and instead expects the user to open the browser and navigate to the address hosting the reports as specified in the CLI. Another important flag is the currency flag (-cur) if unspecified, the currency is set to USD.

### ChatBot

The reporting suite also features a chatbot for querying the analysis in natural language. This feature is still under development, but a version is available in 0.0.4 (not yet released).

![chatbot](https://github.com/supplybi/supplybi.github.io/blob/master/images/dash.gif)

For a more detailed breakdown of the reporting features, please navigate to the [documentation](http://supplychainpy.readthedocs.org/) 

## Docker Image

The docker image for supplychainpy is built from the `continuumio/anaconda3` image, with a pre-installed version of supplychainpy and all the dependencies (see the [dockerfile](https://github.com/KevinFasusi/supplychainpy/blob/master/Dockerfile) for more).

```
    docker run -ti -v directory/on/host:directory/in/container --name fruit-smoothie -p5000:5000 supplychainpy/suchpy bash
```

The port, container name and directories can be changed as needed. Use a shared volume (as shown above) to present a CSV to the container for generating the report.

Make sure you specify the host as "0.0.0.0" for the reporting instance running in the container.

```
    supplychainpy data.csv -a -loc / -lx --host 0.0.0.0
```

## Important Links

- Jupyter Notebooks: [supplychainpy_notebooks](https://github.com/KevinFasusi/supplychainpy_notebooks)
- Documentation: [supplychainpy.readthdocs](http://supplychainpy.readthedocs.org/)
- Website: [supplychainpy.org](http://www.supplychainpy.org/)
- Forum: [google groups](https://groups.google.com/forum/#!forum/supplychainpy)

## License

BSD-3-Clause


## Change Log

### 0.0.5

**Application**

-   **[Bug Fix]** Using Flask's web server for the Dashboard on a public route on a standalone server (--host 0.0.0.0)
-   **[Bug Fix]** Javascript error while loading dashboard.
-   **[New]** Basic ability to run Monte Carlo Simulation and view summarised results in reporting suite.
-   **[Update]** Load scripts use multi-processing for forecast calculations when processing data file.
-   **[Update]** Load scripts using batch process.
-   **[Update]** Debug commandline argument for viewing logging output `--debug'.
-   **[Update]** Use Chat Bot from commandline with `-c` flag. EXPERIMENTAL
-   **[Update]** Recommendation generator takes into account forecasts
-   **[Update]** Flask Blueprints used for reporting views.


**Documentation**

-   **[New]** Wiki started on GitHub for more responsive updates to documentation including changes to source during development.
-   **[Update]** Tutorial.


### 0.0.4 [17 Nov 2016]

Release 0.0.4 has breaking API changes. Namespaces have changed in this release. All the modules previously in the
"demand" package are now inside the "inventory" package. If you have been using the "model_inventory" module, then nothing has
changed, there will not be any break in contracts.

#### Application

-   **[New]** Analytic Hierarchy Process.
-   **[New]** API supports Pandas `DataFrame`.
-   **[New]** Browser based reporting suite, with charts, data summaries and integrated chat bot.
-   **[New]** Dash Bot, a basic chat bot assistant for the data in the reporting suite. Query data using natural language.
-   **[New]** Command line interface for processing .csv to database, launching reports and chat bot.
-   **[New]** "Model_Demand" module containing simple exponential smoothing and holts trend corrected exponential smoothing.
-   **[New]** Summarise and filter your analysis.
-   **[New]** Holts Trend Corrected Exponential Smoothing Forecast and optimised variant (evolutionary algorithm for optimised alpha and gamma)
-   **[New]** Simple Exponential Smoothing (evolutionary algorithm for optimised alpha).
-   **[New]** Evolutionary Algorithms for Smoothing Level Constants (converges on better smoothing levels using genetic algorithm)
-   **[New]** SKU and inventory profile recommendations generator.
-   **[Update]** Explicit internal and public API.
-   **[Update]** Excess, shortages added to the UncertainDemand order_summary.
-   **[Update]** Moved abc_xyz.py, analyse_uncertain_demand.py, economic_order_quantity.py and eoq.pyx from "demand" to "inventory" package
-   **[Update]** "demand" package now contains: evolutionary_algorithms.py, forecast_demand.py and regression.py
-   **[Update]** retail_price added to `model_inventory.analyse_orders`.
-   **[Update]** backlog added to the data format for loading into the analysis.
-   **[Update]** Unit Tests.
-   **[Update]** Docstrings.

#### Documentation

-   **[New]** Reporting Suite Walk Through.
-   **[New]** Declare public API explicitly. describe and document each module and function, give an example also add to website tutorial as Jupyter notebook.
-   **[New]** Docker for supplychainpy quick guide.
-   **[New]** Analytic Hierarchy Process quick guide.
-   **[New]** Inventory Modeling.
-   **[New]** Demand Planning with Pandas.
-   **[Update]** Tutorial.
-   **[Update]** Quick Guide.

### 0.0.3 [30 Mar 2016]

**Application**

-  **[Update]** Compiled Cython (eoq and simulation modules) for OS X, Windows and Linux.
-  **[Update]** Removed z_value, file_type, file_path and reorder_cost parameters from simulate.run_monte_carlo.

**Documentation**

-  **[Update]** Quick Guide


### 0.0.2 [30 Mar 2016]

**Application**

-   **[New]** monte carlo simulation and simulation summary using Cython optimisation.
-   **[New]**  orders analysis optimisation, based on results of the monte carlo simulation.
-   **[New]**  simulate module to api.
-   **[New]**  weighted moving average forecast.
-   **[New]**  moving average forecast.
-   **[New]**  mean absolute deviation.
-   **[Update]** economic order quantity using Cython optimisation.
-   **[Update]** unit tests.

**Documentation**

-   **[New]** Formulas and Equations.
-   **[Update]** Quick Guide.
-   **[Update]** Tutorial.
-   **[Update]** README.md
-   **[Update]** updated data.csv.


## 0.0.1 [20 Feb 2016]

**Application**

-   **[New]** inventory analysis for uncertain demand. Analyse orders from .csv, .txt or from dict.
-   **[New]** inventory analysis summary for uncertain demand. ABC XYZ, economic order quantity (EOQ), reorder level (ROL), demand variability and safety stock.

**Documentation**

-   **[New]** Quick Guide.
-   **[New]** Tutorial.
-   **[New]** Installation.

docs/changelog.rst
Change Log
==========

0.0.5
-----

Application
^^^^^^^^^^^
-   [Bug Fix] Using Flask's web server for the Dashboard on a public route on a standalone server (--host 0.0.0.0)
-   [Bug Fix] Javascript error while loading dashboard.
-   [New Feature] Basic ability to run Monte Carlo Simulation and view summarised results in reporting suite.
-   [Update] Load scripts use multi-processing for forecast calculations when processing data file.
-   [Update] Load scripts using batch process.
-   [Update] Debug commandline argument for viewing logging output `--debug'.
-   [Update] Use Chat Bot from commandline with `-c` flag. EXPERIMENTAL
-   [Update] Recommendation generator takes into account forecasts
-   [Update] Flask Blueprints used for reporting views.


Documentation
^^^^^^^^^^^^^
-   [New] Wiki started on GitHub for more responsive updates to documentation including changes to source during development.
-   [Update] Tutorial.



0.0.4
-----
Release 0.0.4 has breaking API changes. Namespaces have changed in this release. All the modules previously in the
"demand" package are now inside the "inventory" package. If you have been using the "model_inventory" module, then nothing has
changed, there will not be any break in contracts.


Application
^^^^^^^^^^^

-   [Update] Explicit internal and public API.
-   [Update] Excess, shortages added to the UncertainDemand order_summary.
-   [Update] Moved abc_xyz.py, analyse_uncertain_demand.py, economic_order_quantity.py and eoq.pyx from "demand" to "inventory" package
-   [Update] "demand" package now contains: evolutionary_algorithms.py, forecast_demand.py and regression.py
-   [Update] retail_price added to `model_inventory.analyse_orders`.
-   [Update] backlog added to the data format for loading into the analysis.
-   [Update] Unit Tests.
-   [Update] Docstrings.
-   [New Feature] Analytic Hierarchy Process.
-   [New Feature] API supports Pandas `DataFrame`.
-   [New Feature] Browser based reporting suite, with charts, data summaries and integrated chat bot.
-   [New Feature] Dash Bot, a basic chat bot assistant for the data in the reporting suite. Query data using natural language.
-   [New Feature] Command line interface for processing .csv to database, launching reports and chat bot.
-   [New Feature] "Model_Demand" module containing simple exponential smoothing and holts trend corrected exponential smoothing.
-   [New Feature] Summarise and filter your analysis.
-   [New Feature] Holts Trend Corrected Exponential Smoothing Forecast and optimised variant (evolutionary algorithm for optimised alpha and gamma)
-   [New Feature] Simple Exponential Smoothing (evolutionary algorithm for optimised alpha).
-   [New Feature] Evolutionary Algorithms for Smoothing Level Constants (converges on better smoothing levels using genetic algorithm)
-   [New Feature] SKU and inventory profile recommendations generator.

Documentation
^^^^^^^^^^^^^

-   [New] Reporting Suite Walk Through.
-   [Update] Tutorial.
-   [Update] Quick Guide.
-   [New] Declare public API explicitly. describe and document each module and function, give an example also add to website tutorial as Jupyter notebook.
-   [New] Docker for supplychainpy quick guide.
-   [New] Analytic Hierarchy Process quick guide.
-   [New] Inventory Modeling.
-   [New] Demand Planning with Pandas.

0.0.3
-----

Application
^^^^^^^^^^^

-   Compiled Cython (eoq and simulation modules) for OS X, Windows and Linux.
-   Removed z_value, file_type, file_path and reorder_cost parameters from simulate.run_monte_carlo.

Documentation
^^^^^^^^^^^^^

-   Update Quick Guide

0.0.2
-----

Application
^^^^^^^^^^^

-   Added monte carlo simulation and simulation summary using Cython optimisation.
-   Added orders analysis optimisation, based on results of the monte carlo simulation.
-   Added simulate module to api.
-   Added weighted moving average forecast.
-   Added moving average forecast.
-   Added mean absolute deviation.
-   Updated economic order quantity using Cython optimisation.
-   Updated unit tests.

Documentation
^^^^^^^^^^^^^

-   Updates Quick Guide.
-   Updated Tutorial.
-   Updated README.md
-   Added Formulas and Equations.
-   Updated data.csv.

0.0.1
-----

Application
^^^^^^^^^^^

-   Added inventory analysis for uncertain demand. Analyse orders from .csv, .txt or from dict.
-   Added inventory analysis summary for uncertain demand. ABC XYZ, economic order quantity (EOQ), reorder level (ROL),
    demand variability and safety stock.

Documentation
^^^^^^^^^^^^^

-   Added Quick Guide.
-   Added Tutorial.
-   Added Installation.



supplychainpy/scratch.py
# Copyright (c) 2015-2016, The Authors and Contributors
# <see AUTHORS file>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
# following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the
# following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
# following disclaimer in the documentation and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
# products derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import os
import time
from decimal import Decimal
import logging

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.mlab as mlab
from sklearn.datasets import fetch_california_housing
from sklearn.datasets import load_boston
from pandas import Series, DataFrame

# logging.basicConfig(level=logging.CRITICAL, format='%(asctime)s - %(levelname)s - %(message)s')
from supplychainpy import model_inventory
from supplychainpy._csv_management._model._db_setup import transaction_type, metadata
from supplychainpy._helpers._config_file_paths import ABS_FILE_PATH_CSV_MANAGEMENT_CONFIG, \
    ABS_FILE_PATH_APPLICATION_CONFIG
from supplychainpy._helpers._pickle_config import deserialise_config
from supplychainpy.bi._analytical_heirachy_process import _PairwiseComparison
from supplychainpy.demand._forecast_demand import Forecast
from supplychainpy.model_demand import simple_exponential_smoothing_forecast, \
    holts_trend_corrected_exponential_smoothing_forecast_from_file
from supplychainpy.model_inventory import analyse, analyse_orders
from supplychainpy.sample_data.config import ABS_FILE_PATH
from supplychainpy.model_decision import analytical_hierarchy_process


def main():
    pass
    # pp = np.array([[1,2,7,9],[0.5,1,5.0,5.0],[0.1429,0.2,1.0,5.0],[0.1111, 0.2,  0.2, 1.0]])
    # sq = pp*pp
    # np.set_printoptions(precision=3, suppress=True)
    # print(sq)
    #lorry_cost = {'scania': 55000, 'iveco': 79000, 'volvo': 59000, 'navistar': 66000}
    #criteria = ('style', 'reliability', 'comfort', 'fuel_economy')
    #criteria_scores = [(1 / 1, 2 / 1, 7 / 1, 9 / 1), (1 / 2, 1 / 1, 5 / 1, 5 / 1), (1 / 7, 1 / 5, 1 / 1, 5 / 1),
    #                   (1 / 9, 1 / 5, 1 / 5, 1 / 1)]
#
    #options = ('scania', 'iveco', 'navistar', 'volvo' )
    #option_scores = {
    #    'style': [(1 / 1, 1 / 3, 5 / 1, 1 / 5), (3 / 1, 1 / 1, 2 / 1, 3 / 1), (1 / 3, 1 / 5, 1 / 1, 1 / 5),
    #              (5 / 1, 1 / 3, 5 / 1, 1 / 1)],
    #    'reliability': [(1 / 1, 1 / 3, 3 / 1, 1 / 7), (3 / 1, 1 / 1, 5 / 1, 1 / 5), (1 / 3, 1 / 5, 1 / 1, 1 / 5),
    #                    (7 / 1, 5 / 1, 5 / 1, 1 / 1)],
    #    'comfort': [(1 / 1, 5 / 1, 5 / 1, 1 / 7), (1 / 5, 1 / 1, 2 / 1, 1 / 7), (1 / 3, 1 / 5, 1 / 1, 1 / 5),
    #                (7 / 1, 7 / 1, 5 / 1, 1 / 1)],
    #    'fuel_economy': (11, 9, 10, 12)}
    #lorry_decision = analytical_hierarchy_process(criteria=criteria, criteria_scores=criteria_scores, options=options,
    #                                              option_scores=option_scores, quantitative_criteria=('fuel_economy',),
    #                                              item_cost=lorry_cost)
    #print(lorry_decision)
    # metadata.create_all(engine)
    ##print(deserialise_config(ABS_FILE_PATH_CSV_MANAGEMENT_CONFIG),'\n')
    # print(deserialise_config(ABS_FILE_PATH_APPLICATION_CONFIG))
    # inventory_analysis = [i.orders_summary() for i in
    #                      model_inventory.analyse(file_path=ABS_FILE_PATH['COMPLETE_CSV_SM'],
    #                                                                      z_value=Decimal(1.28),
    #                                                                      reorder_cost=Decimal(500),
    #                                                                      file_type='csv')]
    # print(inventory_analysis[0])
    #

    # raw_df = pd.read_csv(ABS_FILE_PATH['COMPLETE_CSV_SM'])
    # # print(raw_df)
    # analysis_df = analyse(df=raw_df, start=1, interval_length=12, interval_type='months')
    # # print(analysis_df[['sku', 'quantity_on_hand', 'excess_stock', 'shortages', 'ABC_XYZ_Classification']])


#
# row_ds = raw_df[raw_df['Sku'] == 'KR202-212'].squeeze()
# print(row_ds[1:12])
# d = simple_exponential_smoothing_forecast(ds=row_ds[1:12], length=12, smoothing_level_constant=0.5)
# print(d)
# htces_forecast = [i for i in holts_trend_corrected_exponential_smoothing_forecast_from_file(
#                           file_path=ABS_FILE_PATH['COMPLETE_CSV_SM'],
#                           file_type='csv',
#                           length=12,
#                           alpha=0.2,
#                           gamma=0.2,
#                           smoothing_level_constant=0.5,
#                           optimise=True)]
#
# for i in htces_forecast:
#    for k in i.values():
#        k.get('standard_error')
# start_time = time.time()
# app_dir = os.path.dirname(__file__, )
# rel_path = 'supplychainpy/data_col.csv'
# abs_file_path = os.path.abspa§th(os.path.join(app_dir, '..', rel_path))

## orders_analysis = model_inventory.analyse_orders_abcxyz_from_file(file_path=abs_file_path,
##                                                                z_value=Decimal(1.28),
###                                                                reorder_cost=Decimal(5000),
##                                                                file_type="csv", length=12)
## for i in orders_analysis:
##    print(i.orders_summary())

## ia = [analysis.orders_summary() for analysis in
##      model_inventory.analyse_orders_abcxyz_from_file(file_path=abs_file_path, z_value=Decimal(1.28),
##                                                      reorder_cost=Decimal(5000), file_type="csv",
##                                                      length=12)]

# orders = [165, 171, 147, 143, 164, 160, 152, 150, 159, 169, 173, 203, 169, 166, 162, 147, 188, 161, 162, 169, 185,
#          188, 200, 229, 189, 218, 185, 199, 210, 193, 211, 208, 216, 218, 264, 304]
#
# total_orders = 0
# avg_orders = 0
#
# for order in orders[:12]:
#    total_orders += order
#
# avg_orders = total_orders / 12
# f = Forecast(orders)
# alpha = [0.2, 0.3, 0.4, 0.5, 0.6]
# s = [i for i in f.simple_exponential_smoothing(*alpha)]
#   p = [i for i in f.holts_trend_corrected_exponential_smoothing(0.5, 0.5, 155.88, 0.8369)]
#   print(p)
# holts_forecast = f.holts_trend_corrected_forecast(forecast=p, forecast_length=4)
# print(holts_forecast)
##sas = holts_trend_corrected_exponential_smoothing_forecast(orders,0.5,0.5,forecast_length=4, initial_period=18,
##                                                           optimise=True)
##print(sas)
##sum_squared_error = f.sum_squared_errors(p, 0.5)
##print(sum_squared_error)
# sum_squared_error = f.sum_squared_errors(s, 0.5)
# standard_error = f.standard_error(sum_squared_error, len(orders), 0.5, 2)
# print(standard_error)
# standard_error = f.standard_error(sum_squared_error, len(orders), 0.5)
#
# evo_mod = OptimiseSmoothingLevelGeneticAlgorithm(orders=orders,
#                                                 average_order=avg_orders,
#                                                 population_size=10,
#                                                 standard_error=standard_error,
#                                                 recombination_type='single_point')
#
# optimal_alpha = evo_mod.initial_population(individual_type='htces')
#
# print(optimal_alpha)
# forecast = [i for i in f.simple_exponential_smoothing(optimal_alpha[1])]
# s = LinearRegression(forecast)
# w = s.least_squared_error()
# mape = f.mean_aboslute_percentage_error_opt(forecast)
# sim = evo_mod.simple_exponential_smoothing_evo(0.5, 12, 'two_point')
# print(sim)
## evo_mod.run_smoothing_level_evolutionary_algorithm(parents=evo_mod.initial_population,

#                                                  standard_error=standard_error,
#                                                  smoothing_level=0.5)

# print([i['orders'].values() for i in ia])
# orders = [i['orders'].values() for i in ia]
# orders1 = [i for i in orders[0]]
# orders2 = [i for i in orders1[0]]
# vector = np.array(orders2)
##print(vector)
# row_vector = vector.reshape((12, 1))
# #print(row_vector)
# column_vector = vector.reshape((1, 12))
# #print(column_vector)
# single_feature_matrix = vector.reshape((1, 12))
# print(single_feature_matrix)
# pd_data = pd.Series(orders2, name='orders')
# print(pd_data)
# mean_expected_value = np.mean(pd_data)

# print(pd_data[0])

# squared_errors_pd = pd.Series(mean_expected_value - pd_data[0])  **2

# print(pd_data)
# print(np.mean(pd_data))
# pd.set_option('display.float_format', lambda x: '%.3f' % x)
# dataset = pd.DataFrame(data=row_vector)
# print(dataset[0].mean())
#
# mean_expected_value = dataset.mean()
#
##squared_error =  pd.Series([mean_expected_value - (x ** 2) for x in dataset[0].astype(float)])
## print(squared_error)
# boston = load_boston()
##california = fetch_california_housing()
# dataset2 = pd.DataFrame(boston.data, columns=boston.feature_names)
# dataset2['target'] = boston.target
# print(dataset2['target'])
##print(dataset2['target'].mean())
# squared_error =  pd.Series(mean_expected_value - dataset['target']) ** 2
# print(squared_error)
# sse = sum(squared_error)
# density_plot = squared_error.plot('hist')
##mean_expected_value = np.mean(dataset2)
##print(mean_expected_value)
#
## launch_report()
# analysis_summary = OrdersAnalysis(analysed_orders=orders_analysis)
# skus = ['KR202-209', 'KR202-210', 'KR202-211']

# skus_description = [summarised for summarised in analysis_summary.describe_sku(*[i['sku'] for i in ia])]
# print(skus_description)

# top_ten_shortages = [item for item in analysis_summary.rank_summary(attribute="shortages", count=10, reverse=True)]

# print(top_ten_shortages)

# simple_forecast  = {analysis.sku_id: analysis.simple_exponential_smoothing_forecast for analysis in
#       model_inventory.analyse_orders_abcxyz_from_file(file_path="data2.csv", z_value=Decimal(1.28),
#                                                       reorder_cost=Decimal(5000), file_type="csv",
#                                                       length=12)}


#
# for p, i in enumerate(simple_forecast, 1):
#    print(simple_forecast.get(i)['forecast'][p-1])

# holts_forecast = {analysis.sku_id: analysis.holts_trend_corrected_forecast for analysis in
#                 model_inventory.analyse_orders_abcxyz_from_file(file_path="data2.csv", z_value=Decimal(1.28),
#                                                                  reorder_cost=Decimal(5000), file_type="csv",
#                                                                length=12)}

# ses_forecast = {analysis.sku_id: analysis.simple_exponential_smoothing_forecast for analysis in
#                 model_inventory.analyse_orders_abcxyz_from_file(file_path="data2.csv", z_value=Decimal(1.28),
#                                                                 reorder_cost=Decimal(5000), file_type="csv",
#                                                                  length=12)}

# for i in holts_forecast:
#   print(holts_forecast.get(i))

# for i in holts_forecast:
#    for g in holts_forecast.get(i)['forecast_breakdown']:
#        print(g)

# for orders in inventory_analysis:
#  print(orders)


#
# summary = OrdersAnalysis(analysed_orders=inventory_analysis)
# abc_raw = summary.abc_xyz_raw
# print(abc_raw.ay)
#
# for analysis in summary.abc_xyz_summary():
#    print(analysis.get("AX"))  # print(analysis.get("AX")["revenue"] for entering cell
#
# ar = [analysis.get("AZ")['revenue'] for analysis in summary.abc_xyz_summary() if analysis.get("AZ")]
# print(ar)
#
# ax_d = abc_raw.ay
#
# for sku in ax_d:
#    print("stuff", sku.get("safety_stock"))
#
# print("AX stuff", ax_d)
#
## for top_shortages in summary.top_sku(attribute="shortages", count=10, reverse=False):
##    print(top_shortages)
## get sku keys from category analysis and unpack_sku_detail for describe sku
# questions = ['KR202-209', 'KR202-210', 'KR202-211']
# for summarised in summary.describe_sku('KR202-217'):
#    print(summarised)
##
# s = [summarised for summarised in summary.describe_sku('KR202-217')]
# print("\n", "new one", s)
##
#
# print("keys {}".format(list(ax_shortages.keys())))
# print("values {}".format(list(ax_shortages.values())))

# top_ten_shortages = [item for item in
#                     SKU(analysed_orders=orders_analysis.orders).top_sku(attribute="shortage", count=10,
#                                                                         reverse=True)]
#
# top_ten_excess = [item for item in
#                  SKU(analysed_orders=orders_analysis.orders).top_sku(attribute="excess_stock", count=10,
#                                                                      reverse=True)]
#
# for order in SKU(analysed_orders=orders_analysis.orders).top_sku(attribute="shortage", count=10, reverse=False):
#    print(order)
#
#
#

# sim_window = simulate.summarize_window(simulation_frame=sim, period_length=12)
#
# sim_frame = simulate.summarise_frame(sim_window)
#
# optimised_orders = simulate.optimise_service_level(service_level=95.0, frame_summary=sim_frame,
#                                                   orders_analysis=orders_analysis.orders, runs=100,
#                                                   percentage_increase=1.30)
#
# sim_frame = simulate.summarise_frame(sim_window)
#
# optimised_orders = simulate.optimise_service_level(service_level=95.0, frame_summary=sim_frame,
#                                                   orders_analysis=orders_analysis.orders, runs=100,
#                                                   percentage_increase=1.30)
# end = time.time()
#
# secs = end - start_time
# print(secs)


if __name__ == '__main__':
    main()
